openssl-sys/build/main.rs

```diff
@@ -82,16 +82,43 @@ fn main() {
     let version = validate_headers(&[include_dir.clone().into()]);
 
     let libs_env = env("OPENSSL_LIBS");
-    let libs = match libs_env.as_ref().and_then(|s| s.to_str()) {
-        Some(ref v) => v.split(":").collect(),
+    let (kind, libs) = match libs_env.as_ref().and_then(|s| s.to_str()) {
+        Some(ref v) => {
+            let libs: Vec<_> = v.split(":").collect();
+            (determine_mode(Path::new(&lib_dir), &libs), libs)
+        }
         None => match version {
-            Version::Openssl10x if target.contains("windows") => vec!["ssleay32", "libeay32"],
-            Version::Openssl11x if target.contains("windows") => vec!["libssl", "libcrypto"],
-            _ => vec!["ssl", "crypto"],
+            Version::Openssl10x if target.contains("windows") => {
+                let libs = vec!["ssleay32", "libeay32"];
+                (determine_mode(Path::new(&lib_dir), &libs), libs)
+            }
+            Version::Openssl11x if target.contains("windows") => {
+                let libs = vec!["libssl", "libcrypto"];
+                if let Ok(kind) = determine_mode(Path::new(&lib_dir), &libs) {
+                    (Ok(kind), libs)
+                } else {
+                    // With mingw (as opposed to msvc) lib*.a files are created if static
```

Comment:
Is the issue here that the mingw can link to either liblibssl.lib or libssl.a?

Reply:
I'll be honest, at this point I'm not sure. Maybe @pyfqm knows? I'll look into it.

Reply:
The [build error](https://ci.appveyor.com/project/sfackler/rust-openssl/builds/32501633/job/sadqhnjnl0jyk5d5) looks like a network problem? It fails to download some file.

Reply:
The build error should be fixed if you rebase.

Reply:
I'll do that right away, thanks!

Reply:
That did it.

Reply:
MinGW linkers threats `.lib` as import libraries, `.a` can be either import or static library when `-Bstatic` is present.

Reply:
This is the core issue:

> Also note that the library names "lib<name>.a" and "lib<name>.lib" are not equivalent; if you have a library named according to the aberrant "lib<name>.lib" convention, it will not be found by an "-l<name>" specification -- if you cannot rename the library, you must use the form "-llib<name>" instead.

http://www.mingw.org/wiki/specify_the_libraries_for_the_linker_to_use

OpenSSL builds produce `libssl.dll` rather than `ssl.dll` which makes the linker unhappy with the normal `-lssl -lcrypto`.

Can this else branch be filtered down to specifically windows-gnu targets? I kind of hate all of this existing logic, but I think it's unfortunately necessary :(

Reply:
@sfackler I think @zonyitoo's explanation in #1256 is good.